File: src/main/java/com/projectswg/forwarder/resources/recording/PacketRecorder.java
package com.projectswg.forwarder.resources.recording;

import java.io.*;
import java.lang.management.ManagementFactory;
import java.lang.management.OperatingSystemMXBean;
import java.time.Instant;
import java.time.ZoneId;
import java.util.Map;
import java.util.Map.Entry;
import java.util.TreeMap;

public class PacketRecorder implements AutoCloseable, Closeable {
	
	private static final byte VERSION = 3;
	
	private final DataOutputStream dataOut;
	
	public PacketRecorder(File file) throws FileNotFoundException {
		dataOut = new DataOutputStream(new FileOutputStream(file));
		writeHeader();
	}
	
	public void close() throws IOException {
		dataOut.close();
	}
	
	public void record(boolean server, byte [] data) {
		synchronized (dataOut) {
			try {
				dataOut.writeBoolean(server);
				dataOut.writeLong(System.currentTimeMillis());
				dataOut.writeShort(data.length);
				dataOut.write(data);
			} catch (IOException e) {
				e.printStackTrace();
			}
		}
	}
	
	private void writeHeader() {
		try {
			dataOut.writeByte(VERSION);
			writeSystemHeader();
		} catch (IOException e) {
			e.printStackTrace();
		}
	}
	
	private void writeSystemHeader() {
		OperatingSystemMXBean os = ManagementFactory.getOperatingSystemMXBean();
		Map<String, String> systemStrings = new TreeMap<>();
		systemStrings.put("os.arch",			os.getArch());
		systemStrings.put("os.details",			os.getName()+":"+os.getVersion());
		systemStrings.put("os.processor_count", Integer.toString(os.getAvailableProcessors()));
		systemStrings.put("java.version",		System.getProperty("java.version"));
		systemStrings.put("java.vendor",		System.getProperty("java.vendor"));
		systemStrings.put("time.time_zone",		ZoneId.systemDefault().getId());
		systemStrings.put("time.current_time",	Instant.now().toString());
		try {
			dataOut.writeByte(systemStrings.size()); // Count of strings
			for (Entry<String, String> e : systemStrings.entrySet())
				dataOut.writeUTF(e.getKey() + "=" + e.getValue());
		} catch (IOException e) {
			e.printStackTrace();
		}
	}
	
}

File: src/main/java/com/projectswg/forwarder/resources/recording/PacketRecorder.kt
package com.projectswg.forwarder.resources.recording

import java.io.*
import java.lang.management.ManagementFactory
import java.time.Instant
import java.time.ZoneId
import java.util.*

class PacketRecorder @Throws(FileNotFoundException::class)
constructor(file: File) : AutoCloseable, Closeable {
	
	private val dataOut: DataOutputStream = DataOutputStream(FileOutputStream(file))
	
	init {
		writeHeader()
	}
	
	@Throws(IOException::class)
	override fun close() {
		dataOut.close()
	}
	
	fun record(server: Boolean, data: ByteArray) {
		synchronized(dataOut) {
			try {
				dataOut.writeBoolean(server)
				dataOut.writeLong(System.currentTimeMillis())
				dataOut.writeShort(data.size)
				dataOut.write(data)
			} catch (e: IOException) {
				e.printStackTrace()
			}
			
		}
	}
	
	private fun writeHeader() {
		try {
			dataOut.writeByte(VERSION.toInt())
			writeSystemHeader()
		} catch (e: IOException) {
			e.printStackTrace()
		}
		
	}
	
	private fun writeSystemHeader() {
		val os = ManagementFactory.getOperatingSystemMXBean()
		val systemStrings = TreeMap<String, String>()
		systemStrings["os.arch"] = os.arch
		systemStrings["os.details"] = os.name + ":" + os.version
		systemStrings["os.processor_count"] = os.availableProcessors.toString()
		systemStrings["java.version"] = System.getProperty("java.version")
		systemStrings["java.vendor"] = System.getProperty("java.vendor")
		systemStrings["time.time_zone"] = ZoneId.systemDefault().id
		systemStrings["time.current_time"] = Instant.now().toString()
		try {
			dataOut.writeByte(systemStrings.size) // Count of strings
			for ((key, value) in systemStrings)
				dataOut.writeUTF("$key=$value")
		} catch (e: IOException) {
			e.printStackTrace()
		}
		
	}
	
	companion object {
		
		private const val VERSION: Byte = 3
		
	}
	
}
